File: solutions/elixir/bob/1/lib/bob.ex
defmodule Bob do
  @response %{
    whatever: "Whatever.",
    chill: "Whoa, chill out!",
    sure: "Sure.",
    calm: "Calm down, I know what I\'m doing!",
    fine: "Fine. Be that way!"
  }

  @spec hey(String.t()) :: String.t()
  def hey(input) do
    cleaned =
      input
      |> String.trim()

    cond do
      String.length(cleaned) == 0 -> @response.fine
      question?(cleaned) and shouting?(cleaned) -> @response.calm
      question?(cleaned) -> @response.sure
      shouting?(cleaned) -> @response.chill
      true -> @response.whatever
    end
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end

  def shouting?(input) do
    String.upcase(input) == input &&
      String.match?(input, ~r/[[:alpha:]]/ui)
  end
end

File: solutions/javascript/bob/1/bob.js
const response = {
  whatever: 'Whatever.',
  chill: 'Whoa, chill out!',
  sure: 'Sure.',
  calm: 'Calm down, I know what I\'m doing!',
  fine: 'Fine. Be that way!'
}

const question = (message) => message.endsWith('?')

const shouting = (message) => {
  const cleaned = message.replace(/[^a-z]/gi, '')
  return ((cleaned > '') && (cleaned.toUpperCase() === cleaned))
}

export const hey = (message) => {
  message = message.trim()
  if (message === '') return response.fine
  if (question(message) && shouting(message)) return response.calm
  if (question(message)) return response.sure
  if (shouting(message)) return response.chill
  return response.whatever
}
